### PR TITLE
Stage6

### PR DESCRIPTION
**Pull request by KateMoreva**

### CODE REVIEW

**Comment by KateMoreva** [comment]
В реализации была допущена ошибка(и), из-за которой не получилось достичь выигрыша по времени при обработке запросов, и которая вызывает ConcurrentModificationException: null при первых N запросах к серверу. 

Ошибка будет найдена и исправлена в ближайшее время.

Результаты тестирования данной реализации:

### PUT запрсы

-t64 -c64 -d60s -R2000 -s ./wrk/put.lua --latency http://localhost:8080

**CPU**


**Alloc**


**Lock**



Running 1m test @ http://localhost:8080
  64 threads and 64 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.83ms    0.89ms  17.30ms   88.28%
    Req/Sec    33.03     48.83   111.00     68.54%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%    1.70ms
 75.000%    2.05ms
 90.000%    2.54ms
 99.000%    5.45ms
 99.900%   10.94ms
 99.990%   15.59ms
 99.999%   16.62ms
100.000%   17.31ms
  120027 requests in 1.00m, 7.67MB read
Requests/sec:   2001.50
Transfer/sec:    130.96KB

### GET запросы

-t64 -c64 -d60s -R2000 -s ./wrk/get.lua --latency http://localhost:8080

**CPU**


**Alloc**


**Lock**


Running 1m test @ http://localhost:8080
  64 threads and 64 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.77ms    0.91ms  18.94ms   94.81%
    Req/Sec    33.06     48.89   111.00     68.53%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%    1.66ms
 75.000%    1.96ms
 90.000%    2.23ms
 99.000%    5.76ms
 99.900%   13.20ms
 99.990%   16.75ms
 99.999%   18.69ms
100.000%   18.96ms
  120008 requests in 1.00m, 8.06MB read
Requests/sec:   2002.26
Transfer/sec:    137.67KB

**Comment by KateMoreva** [comment]
Нагрузочное тестирование было проведено повторно, тк  реализация претерпела изменения. Так, например, было убрано повторное хождение в dao при get запросе, которое было видно на предыдущих flame graph. 

### PUT

-t64 -c64 -d60s -R2000 -s ./wrk/put.lua --latency http://localhost:8080

**CPU**

  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     2.86ms    1.53ms  18.22ms   78.44%
    Req/Sec    32.47     44.22   111.00     76.40%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%    2.42ms
 75.000%    3.73ms
 90.000%    4.65ms
 99.000%    8.26ms
 99.900%   13.84ms
 99.990%   17.06ms
 99.999%   17.95ms
100.000%   18.24ms
  120030 requests in 1.00m, 7.67MB read
Requests/sec:   2001.29
Transfer/sec:    130.94KB

Отправка ответа 3.47%
replication 5.04%
DAO.upsert 1.71%

**Alloc**

  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     6.52ms    2.65ms  22.46ms   75.54%
    Req/Sec    32.04     35.27   133.00     84.00%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%    6.69ms
 75.000%    7.82ms
 90.000%    9.00ms
 99.000%   14.71ms
 99.900%   18.98ms
 99.990%   21.39ms
 99.999%   22.43ms
100.000%   22.48ms
  120019 requests in 1.00m, 7.67MB read
Requests/sec:   2001.41
Transfer/sec:    130.95KB

Работа селекторов 37.6%
Отправка ответа 5.28%
DAO.upsert 6.83%

**Lock**

  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     6.95ms    2.68ms  23.31ms   75.47%
    Req/Sec    31.99     31.69   117.00     91.12%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%    7.09ms
 75.000%    8.09ms
 90.000%    9.51ms
 99.000%   15.10ms
 99.900%   18.35ms
 99.990%   22.70ms
 99.999%   23.31ms
100.000%   23.33ms
  120018 requests in 1.00m, 7.67MB read
Requests/sec:   2002.36
Transfer/sec:    131.01KB

Блокировка CompletableFuture занимает 30%
6% держит executor service 
Планировщик client 45.58%
16.03% client


### GET

-t64 -c64 -d60s -R2000 -s ./wrk/get.lua --latency http://localhost:8080

**CPU**

  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     3.04ms    1.89ms  23.18ms   86.22%
    Req/Sec    32.77     47.16   111.00     67.21%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%    2.53ms
 75.000%    3.72ms
 90.000%    5.09ms
 99.000%   10.90ms
 99.900%   17.89ms
 99.990%   21.47ms
 99.999%   22.93ms
100.000%   23.20ms
  120027 requests in 1.00m, 8.06MB read
Requests/sec:   2002.94
Transfer/sec:    137.72KB

Работа селекторов 8.84%
Отправка ответа 6.65%
DAO.getCell 2.69%
replication 5.73%

**Alloc**

  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     3.70ms    2.13ms  22.00ms   76.69%
    Req/Sec    32.55     45.31   111.00     69.37%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%    3.64ms
 75.000%    4.90ms
 90.000%    5.83ms
 99.000%   11.72ms
 99.900%   15.53ms
 99.990%   18.01ms
 99.999%   18.85ms
100.000%   22.01ms
  120041 requests in 1.00m, 8.06MB read
Requests/sec:   2002.10
Transfer/sec:    137.66KB

Работа селекторов 16.47%
Отправка ответа 7.28%
DAO.getCell 7.23%
replication 19.52%

**Lock** 

  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     3.82ms    2.24ms  26.16ms   77.85%
    Req/Sec    32.54     44.45   111.00     73.18%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%    3.80ms
 75.000%    5.10ms
 90.000%    6.00ms
 99.000%   11.77ms
 99.900%   18.43ms
 99.990%   24.38ms
 99.999%   25.95ms
100.000%   26.17ms
  120029 requests in 1.00m, 8.06MB read
Requests/sec:   2002.12
Transfer/sec:    137.66KB

Блокировка локального запроса занимает 28.11%
7.60% держит executor service 
Планировщик client 40%
22.94% client

Блокировка локального запроса занимает 28.11% 7.60% держит executor service Планировщик client 40% 22.94% client
Ухудшения по количеству обработанных заявок, по сравнению с предыдущем этапом, нет.
Среднее время ответов на get запросы возросло примерно в 1.6 раз, на put запросы в 1.2 раза. Stdev стало немного лучше.

Полные результаты приведены [тут](https://github.com/polis-mail-ru/2020-highload-dht/blob/699a920151fd29a16e61f2d3939fc4e108e1532e/tests/stage6/results2.md)

File: src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+
+    private List<CompletableFuture<ResponseValue>> replication(final MySimpleHttpServer.Action action,
+                                                               final ByteBuffer key,
+                                                               final Topology<String> topology,
+                                                               final Context context) {
+        Set<String> nodes = Collections.emptySet();
+        try {
+            nodes = topology.primaryFor(key, context.getReplicaFactor(), context.getReplicaFactor().getAck());
+        } catch (IllegalArgumentException e) {
+            requestHelper.sendLoggedResponse(context.getSession(), new Response(Response.BAD_REQUEST, Response.EMPTY));
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java:204]: Давайте явно `return` напишем? Тогда не придется дефолтное значение для `nodes` задавать.

File: src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        return resultsFuture;
+    }
+
+    /**
+     * Collects responses.
+     */
+    public CompletableFuture<List<ResponseValue>> collect(final List<CompletableFuture<ResponseValue>> responseValues,
+                                                          final int ack, final Executor clientExecutor) {
+        final AtomicInteger numberOfErrors = new AtomicInteger(-1);
+        final List<ResponseValue> results = new ArrayList<>();
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java:125]: Ой-ей. Вы пытаетесь работать с непотокобезопасной коллекцией из асинхронного кода (из лямбды). Прямой путь к конкарренси багам.

File: src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    public boolean isProxied(final Request request) {
+        return request.getHeader(PROXY_HEADER) != null;
+    }
+
+    /**
+     * Takes timestamp from header or -1 if null.
+     */
+    public long getTimestamp(final Response response) {
+        final String timestamp = response.getHeader(TIMESTAMP);
+        return timestamp == null ? -1 : Long.parseLong(timestamp);
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java:185]: Может не распарситься и вылететь `NumberFormatException`. Лучше сделать бронейбоно и выловить исключение.

File: src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+        }
+        final List<CompletableFuture<ResponseValue>> results = new ArrayList<>(nodes.size());
+        for (final String node : nodes) {
+            if (topology.isMe(node)) {
+                getLocalResults(action, results);
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java:209]: Почему бы не вернуть `CompletableFuture` из `getLocalResults`?

File: src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+                                                               final ByteBuffer key,
+                                                               final Topology<String> topology,
+                                                               final Context context) {
+        Set<String> nodes = Collections.emptySet();
+        try {
+            nodes = topology.primaryFor(key, context.getReplicaFactor(), context.getReplicaFactor().getAck());
+        } catch (IllegalArgumentException e) {
+            requestHelper.sendLoggedResponse(context.getSession(), new Response(Response.BAD_REQUEST, Response.EMPTY));
+        }
+        final List<CompletableFuture<ResponseValue>> results = new ArrayList<>(nodes.size());
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java:206]: Если уж задавать дефолтное значение, то тут это скорее `Replicas.from`.

File: src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+        } catch (RejectedExecutionException e) {
+            requestHelper.sendLoggedResponse(session, new Response(Response.SERVICE_UNAVAILABLE, Response.EMPTY));
+        }
+    }
+
+    private void defineMethod(final Request request, final HttpSession session, final ByteBuffer key,
+                              final Replicas replicasFactor, final boolean isProxy) {
+        final Context context = new Context(session, isProxy, request, replicasFactor);
+        CompletableFuture.runAsync(() -> {
+        switch (request.getMethod()) {
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java:154]: Отступы съехали.

File: src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+        });
+        return resultsFuture;
+    }
+
+    /**
+     * Collects responses.
+     */
+    public CompletableFuture<List<ResponseValue>> collect(final List<CompletableFuture<ResponseValue>> responseValues,
+                                                          final int ack, final Executor clientExecutor) {
+        final AtomicInteger numberOfErrors = new AtomicInteger(-1);
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java:124]: Логичнее, наверное, с нуля начать.

File: src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+    private static final String TIMESTAMP = "Timestamp";
+    static final String PROXY_HEADER_KEY = "X-Proxy";
+    private static final Logger log = LoggerFactory.getLogger(MySimpleHttpServer.class);
+    private final Duration timeout = Duration.ofSeconds(1);
+    private final ExecutorService executorService;
+    private final Executor clientExecutor;
+    private final Topology<String> topology;
+    private final MyRequestHelper requestHelper;
+    private final Replicas quorum;
+    private final java.net.http.HttpClient client;
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java:54]: Почему fully qualified name?

File: src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+            log.error("DELETE method failed on /v0/entity for id {}.", key.get(), e);
+            return new Response(Response.INTERNAL_ERROR, Response.EMPTY);
+        }
+    }
+
+    /**
+     * Merges responses for GET request.
+     */
+    public CompletableFuture<ResponseValue> merge(final CompletableFuture<List<ResponseValue>> future) {
+        final CompletableFuture<ResponseValue> resultsFuture = new CompletableFuture<>();
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java:96]: Можно использовать `future.thenApply()` и не создавать самостоятельно новую фьючу. И лучше внятное название дать, что за фьюча (`responses` или типа того).

File: src/main/java/ru/mail/polis/service/kate/moreva/Replicas.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+        return new Replicas(n, count);
+    }
+
+    /**
+     * Parses the request to get the needed number of answers (ack) and nodes (from).
+     * */
+    public static Replicas parser(final String replicas) {
+        final List<String> params = Splitter.on('/').splitToList(replicas);
+        final int ack = Integer.parseInt(params.get(0));
+        final int from = Integer.parseInt(params.get(1));
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/Replicas.java:27]: Можем получить `IndexOutOfBoundException`.

**Review by vaddya — CHANGES_REQUESTED** [review]
Замечания по реализации.

File: src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java (hunk 8 above)
Comment (KateMoreva) [reply src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java]: Переводила на java.net.http.HttpClient постепенно, забыла поправить после удаления клиента из one.nio.

**Comment by KateMoreva** [comment]
> Замечания по реализации.

Спасибо за замечания, все поправила.

File: src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+        }).exceptionally(e -> {
+            log.error("Error while merge ", e);
+            return null;
+        });
+    }
+
+    /**
+     * Collects responses.
+     */
+    public CompletableFuture<List<ResponseValue>> collect(final List<CompletableFuture<ResponseValue>> responseValues,
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java:117]: Какое поведение будет, если `responseValues.size() < ack`?

File: src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+     * {@code 201} (new data created).
+     * {@code 202} (data deleted).
+     * {@code 405} (unexpected method).
+     * {@code 500} (internal server error occurred).
+     */
+    @Path("/v0/entity")
+    public void entity(@Param(value = "id", required = true) final String id, final Request request,
+                       final HttpSession session, @Param("replicas") final String replicas) {
+        try {
+            if (id.isBlank()) {
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java:128]: Логичнее будет в блоке `try` оставить только вызов `execute()`.

File: src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+            final ByteBuffer key = ByteBuffer.wrap(id.getBytes(Charsets.UTF_8));
+            defineMethod(request, session, key, replicasFactor, isProxy);
+        } catch (IllegalArgumentException e) {
+            requestHelper.sendLoggedResponse(session, new Response(Response.BAD_REQUEST, Response.EMPTY));
+        }
+    }
+
+    private void defineMethod(final Request request, final HttpSession session, final ByteBuffer key,
+                              final Replicas replicasFactor, final boolean isProxy) {
+        final Context context = new Context(session, isProxy, request, replicasFactor);
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java:161]: Может создать контекст уровнем выше? Тогда в этот метод можно будет передавать 1-2 параметра.

File: src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+                    break;
+                default:
+                    log.error("Not allowed method on /v0/entity");
+                    requestHelper.sendLoggedResponse(session,
+                            new Response(Response.METHOD_NOT_ALLOWED, Response.EMPTY));
+                    break;
+            }
+        }, clientExecutor).exceptionally(e -> {
+            log.error("Error while executing method ", e);
+            requestHelper.sendLoggedResponse(context.getSession(),
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java:181]: Мы точно не отправим в сессию больше одного ответа? Обработка ошибок расползлась по всему классу, вижу минмиум два блока `exceptionally` и несколько `try-catch`, из которых что-то отправляется в сессию.

File: src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
+                    new Response(Response.GATEWAY_TIMEOUT, Response.EMPTY));
+            return null;
+        });
+    }
+
+    void executeMethod(final Context context, final ByteBuffer key, final Action action) {
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java:187]: `private`?

File: src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java [hunk 16]
```diff
@@ -0,0 +1,277 @@
...
+                }
+            }
+        } catch (IllegalArgumentException e) {
+            requestHelper.sendLoggedResponse(context.getSession(), new Response(Response.BAD_REQUEST, Response.EMPTY));
+        }
+        return results;
+    }
+
+    private CompletableFuture<ResponseValue> getLocalResults(final Action action) {
+        return CompletableFuture.supplyAsync(() -> {
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java:231]: А на каком потоке будет выполняться этот код?

**Review by vaddya — CHANGES_REQUESTED** [review]
Еще немного.

File: src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java (hunk 11 above)
Comment (KateMoreva) [reply src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java:117]: В этом случае collect отдаст пустой List, merge прокинет null в executeMethod, где в блоке exceptionally будет оправлено 504.

File: src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java (hunk 14 above)
Comment (KateMoreva) [reply src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java]: Не должны, отправка ответов в блоке обработки IllegalArgumentException не будет произведена, тк условия, при которых можно получить эту ошибку были проверяются и обрабатываются при получении запроса.

File: src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java [hunk 17]
```diff
@@ -0,0 +1,272 @@
...
+                    results.add(result);
+                }
+            }
+        } catch (IllegalArgumentException e) {
+            requestHelper.sendLoggedResponse(context.getSession(), new Response(Response.BAD_REQUEST, Response.EMPTY));
+        }
+        return results;
+    }
+
+    private CompletableFuture<ResponseValue> getLocalResults(final Action action) {
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java:225]: Теперь получается, что вместо того, чтобы зашедулить поиск в локальном DAO и сделать запросы к другим нодам, мы сначала синхронно будем искать в DAO (хотя могли бы отправить запросы, а пока ждем ответы искать локально). Нужно вернуть асинхронность, но явно указать тред пул, на котором выполняться. В вашем прошлом варианте он не был указан, а значит использовался `ForkJoinPool#commonPool()` вместо нашего.

File: src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java (hunk 11 above)
Comment (vaddya) [reply src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java:117]: Вы уверены? У меня есть ощущение, что если, например, `responseValues.size() == 1`, а `ack = 3` (то есть ждем 3 подтверждений, хотя передаем сюда только 1 фьючу), то `resultsFuture` никогда ни выполнится, ни зафейлится.

Comment (vaddya) [reply src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java:117]: Это я все к тому, что лучше явно провалидировать входные параметры и бросить исключение в случае чего.

Comment (KateMoreva) [reply src/main/java/ru/mail/polis/service/kate/moreva/MyRequestHelper.java:117]: Вы правы, лучше пробросить исключение, исправила.

File: src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java (hunk 17 above)
Comment (KateMoreva) [reply src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java]: Поняла, спасибо, исправила.

File: tests/stage6/results2.md [hunk 18]
```diff
@@ -0,0 +1,1114 @@
...
+Transfer/sec:    137.66KB
+
+Блокировка локального запроса занимает 28.11%
+7.60% держит executor service 
+Планировщик client 40%
+22.94% client
+
+Ухудшения по количеству обработанных заявок, по сравнению с предыдущем этапом, нет.
+
+Среднее время ответов на get запросы возросло примерно в 1.6 раз, на put запросы в 1.2 раза. Stdev стало немного лучше.
```
Comment (vaddya) [inline tests/stage6/results2.md:1114]: То есть при переходе от one-nio клиента к асинхронному, время так сильно выросло? :(

Comment (KateMoreva) [reply tests/stage6/results2.md:1114]: К сожалению, да. Правда, тестирование было до последних изменений, но они не должны радикально улучшить ситуацию.

Comment (vaddya) [reply tests/stage6/results2.md:1114]: Интересно, с чем это связано может быть. Раньше, например, при `2/3` всегда ждали 3 ответов, а теперь 2 самых быстрых. Наверное, в рамках одной машины это просто незаметно.

**Review by vaddya — APPROVED** [review]
Принято.

**Pull request closed without merging**